service/storage_service.cc

```diff
@@ -6178,14 +6178,8 @@ future<> storage_service::await_topology_quiesced() {
         co_return;
     }
 
```

Comment:
> and if it's false it means the topology is no longer busy.

I already complained once that it is not cleat to me what this function is actually tries to do, but the definition above is incorrect. If it is false it means that the topology was not busy at least once after the function was called. What is this useful for I cannot imagine.

Reply:
> the definition above is incorrect. If it is false it means that the topology was not busy at least once after the function was called.

Of course the condition can change again as soon as we notice it, but you can say that about every function which waits for a predicate. This is still useful because if the topology became not busy at some point it means that the backlog created by earlier events was cleared at some point. For example, after replacing a node, replicas were rebuilt.

Reply:
> > the definition above is incorrect. If it is false it means that the topology was not busy at least once after the function was called.
> 
> Of course the condition can change again as soon as we notice it, but you can say that about every function which waits for a predicate. 

There are functions that take a lock and guaranty that a predicate is true after returning. There are cases where there is no concurrency and then the condition is true until next  yield. Those are not the cases here of course.

> This is still useful because if the topology became not busy at some point it means that the backlog created by earlier events was cleared at some point. For example, after replacing a node, replicas were rebuilt.

True, but you need to make sure that the request started to execute. I am not sure we can guaranty that the coordinator never becomes non busy until all request are processed. It may become non busy between requests I think, but if not is it an implementation detail we want to rely on?

Reply:
It's unclear why repair will ever execute on a busy system. Individual tablet movements don't block each other, while a single tablet movement block repair. The waiting is done without any kind of a pending lock on the topology that would prevent new tablet movements from starting. 

It's important to achieve some kind of FIFO between tablet movements and repair. I guess it's in the plan, so this comment is not a review request, it's a follow up request and question.

Reply:
> It's unclear why repair will ever execute on a busy system. Individual tablet movements don't block each other, while a single tablet movement blocks repair. The waiting is done without any kind of a pending lock on the topology that would prevent new tablet movements from starting.
> 
> It's important to achieve some kind of FIFO between tablet movements and repair. I guess it's in the plan, so this comment is not a review request, it's a follow up request and question.

Yes, there is no preemption, so rebalancing will block repair. This will be addressed when we implement the original design to track repair requests in group0 and schedule tablet repair from tablet schduler.

Reply:
> True, but you need to make sure that the request started to execute. I am not sure we can guaranty that the coordinator never becomes non busy until all request are processed. It may become non busy between requests I think, but if not is it an implementation detail we want to rely on?

Note that this patch doesn't rely on this. But I think we should guarantee it one way or the other. We could invoke the load-balancer to check if !busy topology has backlog. But it seems expensive, so guaranteeing in the coordinator that !busy topology implies no load-balancer backlog may be better, and not that hard (just call the load balancer on the exit edges).

Reply:
How expensive is to check the backlog? If we want some kind of guaranty from the coordinator we need to document it and make it somehow hard to break without noticing.

Reply:
> How expensive is to check the backlog? 

It's O(node_count + tablet_replica_count) if no balancing is needed. Not the end of the world, depending on how frequently we check busy(). And assuming that load balancer will stay stateless, so that it can be called not only by the coordinator but from any node and shard.